common/src/main/java/io/pravega/common/util/btree/BTreePage.java

```diff
@@ -0,0 +1,788 @@
+    void update(@NonNull Collection<PageEntry> entries) {
+        if (entries.isEmpty()) {
+            // Nothing to do.
+            return;
+        }
+
+        // Apply the in-place updates and collect the new entries to be added.
+        val newEntries = applyUpdates(entries);
+        if (newEntries.isEmpty()) {
+            // Nothing else to change. We've already updated the keys in-place.
+            return;
+        }
+
+        val newPage = insertNewEntries(newEntries);
+
+        // Make sure we swap all the segments with those from the new page. We need to release all pointers to our
+        // existing buffers.
+        this.header = newPage.header;
+        this.data = newPage.data;
+        this.contents = newPage.contents;
+        this.footer = newPage.footer;
+        this.count = newPage.count;
+    }
+
+    /**
+     * Updates the contents of this BTreePage so that it does not contain any entry with the given Keys anymore.
+     *
+     * After this method completes, this BTreePage:
+     * * May underflow (a merge may be necessary)
+     * * Will have all entries sorted by Key
+     * * Will reuse the same underlying buffer as before (no new buffers allocated). As such it may underutilize the buffer.
+     *
+     * @param keys A Collection of Keys to remove. The Keys need not be sorted.
+     */
+    void delete(@NonNull Collection<ByteArraySegment> keys) {
+        if (keys.isEmpty()) {
+            // Nothing to do.
+            return;
+        }
+
+        // Locate the positions of the Keys to remove.
+        val removedPositions = searchPositions(keys);
+        if (removedPositions.size() > 0) {
+            // Remove them.
+            removePositions(removedPositions);
+        }
+    }
+
+    /**
+     * Gets a ByteArraySegment representing the value mapped to the given Key.
+     *
+     * @param key The Key to search.
+     * @return A ByteArraySegment mapped to the given Key, or null if the Key does not exist. Note that this is a view
+     * inside a larger array and any modifications to that array will be reflected in this. If this value needs to be held
+     * for longer then it is recommended to get a copy of it (use getCopy()).
+     */
+    ByteArraySegment searchExact(@NonNull ByteArraySegment key) {
+        val pos = search(key, 0);
+        if (!pos.isExactMatch()) {
+            // Nothing found.
+            return null;
+        }
+
+        return getValueAt(pos.getPosition());
+    }
+
+    /**
+     * Performs a (binary) search for the given Key in this BTreePage and returns its position.
+     *
+     * @param key      A ByteArraySegment that represents the key to search.
+     * @param startPos The starting position (not array index) to begin the search at. Any positions prior to this one
+     *                 will be ignored.
+     * @return A SearchResult instance with the result of the search.
+     */
+    SearchResult search(@NonNull ByteArraySegment key, int startPos) {
```

Comment:
I think you can probably generalize CollectionHelpers.binarySearch() to work here.